---
## Hausaufgabe 4 - Verbesserungsverfahren
---

### Achtung: Bitte überprüfen Sie vor Abgabe der Hausaufgabe, ob das Notebook richtig gespeichert wurde. 
Die Speicherung von Notebooks funktioniert über das "Disketten"-Symbol im Notebook oder über die Shortcuts Strg + S (Win) und CMD + S (MAC).

---
## Bewertung
---

#### Erreichbare Punkte: 25
---
#### Erreichte Punkte:   -
---
---
Aufgabe 1:<br>
<br>
Aufgabe 2:<br>
<br>
Aufgabe 3:<br>

In [16]:
permutation = list(job for job in range(6))

def blockSwap(initial_permutation, i, j, k):
    '''Swaps the blocks permutation [i+k] and puts it at position j'''
    new_permutation = []
    for pos in range(len(initial_permutation)):
        if pos in range(i, i+k):
            continue
        new_permutation.append(permutation[pos])
    for idx, pos in enumerate(range(i, i+k)):
        new_permutation.insert(j+idx, initial_permutation[pos])
    return new_permutation

print(permutation)
print(blockSwap(permutation, 0, 3, 3))

[0, 1, 2, 3, 4, 5]
[3, 4, 5, 0, 1, 2]


In [19]:
def blockSwap2(initial_permutation, i, j, k):
    '''Swaps the blocks permutation [i+k] and puts it at position j'''
    block = initial_permutation[i:i+k]
    rest = initial_permutation[:i] + initial_permutation[i+k:]
    new_permutation = rest[:j] + block + rest[j:]
    return new_permutation

print(permutation)
print(blockSwap2(permutation, 3, 0, 3))

[0, 1, 2, 3, 4, 5]
[3, 4, 5, 0, 1, 2]


#### Aufgabe 1 - Theorie: Block Move (3 Punkte)

Ein Block Move $M(\text{block}, s, i, j, k)$ beschreibt eine Verschiebung von $k$ aufeinanderfolgenden Elementen. Die Elemente an Position $i$ bis einschließlich $i + k - 1$ werden entnommen und an Position $j$ neu eingefügt.

Bestimmen Sie die Komplexität der Block-Move Nachbarschaft für $k = 1$ und $k = 3$ und die ganze Nachbarschaftsfamilie, d.h. für alle $k$. Nehmen Sie an, dass eine Lösung durch eine Permutation der Länge $n$ repräsentiert ist.

$$ |N(\text{block}, s, k = 1)| = ? $$
$$ |N(\text{block}, s, k = 3)| = ? $$
$$ \sum_{k = 1}^{n}|N(\text{block}, s, k)| = ? $$

**Hinweise:** Geben Sie die Komplexität in Abhängigkeit von $n$ und $k$ an.

**Literaturverweis** (in OPAL vorhanden): Xu, H., Lü, Z., & Cheng, T. C. E. (2014). Iterated local search for single-machine scheduling with sequence-dependent setup times to minimize total weighted tardiness. Journal of Scheduling, 17(3), 271-287.

**Tipp:**
<details>

Wie viele Blöcke sind in einer Sequenz enthalten? An wieviele Positionen kann ein Block platziert werden? Wieviele Dopplungen/Symmetrien gibt es?
</details>

#### k=1
- mögliche Anzahl an Blöcke = $n$
- mögliche Anzahl an Positionen = $n-1$     (eigene Position fällt weg)  
$$\Longrightarrow |N(block, s, k=1)| = n(n-1)$$

Bei $k=1$ handelt es sich um die *Insertion* Nachbarschaft. Es treten Dopplungen auf, bspw. $(i=1, j=2)$ und $(i=2, j=1)$. Formalisiert heißt das $j=i-1$ oder $i=j+1$ ist eine Dopplung. Somit ergibt sich vereinfacht $(n-1)^2$.

#### k=3
- mögliche Anzahl an Blöcken = $n-k+1$
- mögliche Anzahl an Positionen = $n-k$     (Positionen der Elemente in $i+k$ entfallen)  
$$\Longrightarrow |N(block, s, k=3)| = (n-k+1)(n-k)$$

Bei $k>2$ können in Abhängigkeit von der Anzahl $n$ ebenfalls Dopplungen auftreten. Im Wesentlichen kommt es zu Dopplung, wenn benachbarte Blöcke getauscht werden, d.h. bei $k=2$ bspw. $(i=0,j=2)$ und $(i=2,j=0)$. Formal zusammengetragen lässt sich die Anzahl der Dopplungen in Abhängigkeit von Blöcklänge $k$ und Elementanzahl $n$ durch $max(0, n-2k+1)$ abbilden. Somit is zusammengefasst:  
$$\Longrightarrow |N(block, s, k=3)| = (n-k+1)(n-k)-max(0,n-6+1)$$

#### Gesamte Nachbarschaftsfamilie
Zusammengefasst für alle Nachbarschaften, d.h. alle Blocklängen lässt es sich wie folgt zusammenfassen:  
$$\Longrightarrow |N(block, s, k)| = \sum_{k=1}^{n-1} (n-k+1)(n-k)-max(0,n-2k+1)$$  
($n$ kann in Summe vernachlässigt werden, da bei $k=n$ null Tausche infrage kommen)

#### Aufgabe 2 - Implementierung I: Block-Move (18 Punkte)

Erweitern Sie den in der Lehrveranstaltung erstellten Solver um eine Block-Move Nachbarschaft für $k = 3$. Erstellen Sie die Klassen `BlockMoveK3` und `BlockNeighborhoodK3`.



**Hinweise:** 

Implementieren Sie ihre Lösung in den bereitgestellten `.py` Dateien aus der Veranstaltung, nicht hier im Notebook. Bitte kopieren Sie der Übersichtlichkeit wegen trotzdem Ihre in der `.py`-Datei implementierten Block-Move Komponenten hier ins Notebook. Geben Sie alle zum fehlerfreien Ausführen notwendigen Dateien ab.


Testen Sie die Nachbarschaft auf dem Datensatz `InputFlowshopSIST20.json`. Verwenden Sie die BestImprovement-Strategie und die FCFS-Startlösung. Mögliches erwartetes Ergebnis:


In [1]:
from Solver import *

data = InputData("../../../data/Prescriptive_Homework/InputFlowshopSIST20.json")

localSearch = IterativeImprovement(data, 'BestImprovement', ['BlockK3'])

solver = Solver(data, 1008)

solver.RunLocalSearch(
    constructiveSolutionMethod='FCFS',
    algorithm=localSearch)

Generating an initial solution according to FCFS.
Constructive solution found.
The permutation [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1482
The permutation [0, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 1, 2, 3, 19] results in a Makespan of 1356
The permutation [17, 18, 1, 0, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 2, 3, 19] results in a Makespan of 1289
The permutation [17, 4, 18, 1, 0, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 2, 3, 19] results in a Makespan of 1267
The permutation [8, 9, 10, 17, 4, 18, 1, 0, 5, 6, 7, 11, 12, 13, 14, 15, 16, 2, 3, 19] results in a Makespan of 1249
The permutation [8, 10, 17, 4, 9, 18, 1, 0, 5, 6, 7, 11, 12, 13, 14, 15, 16, 2, 3, 19] results in a Makespan of 1231
The permutation [8, 10, 17, 4, 9, 18, 1, 0, 5, 6, 7, 11, 12, 13, 14, 15, 19, 16, 2, 3] results in a Makespan of 1224
Best found Solution.
The permutation [8, 10, 17, 4, 9, 18, 1, 0, 5, 6, 7, 11, 12, 13, 14, 15, 19, 16, 

Möglicher Output:

    Generating an initial solution according to FCFS.
    Constructive solution found.
    The permutation [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1482
    The permutation [0, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 1, 2, 3, 19] results in a Makespan of 1356
    The permutation [17, 18, 1, 0, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 2, 3, 19] results in a Makespan of 1289
    The permutation [17, 4, 18, 1, 0, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 2, 3, 19] results in a Makespan of 1267
    The permutation [8, 9, 10, 17, 4, 18, 1, 0, 5, 6, 7, 11, 12, 13, 14, 15, 16, 2, 3, 19] results in a Makespan of 1249
    The permutation [8, 10, 17, 4, 9, 18, 1, 0, 5, 6, 7, 11, 12, 13, 14, 15, 16, 2, 3, 19] results in a Makespan of 1231
    The permutation [8, 10, 17, 4, 9, 18, 1, 0, 5, 6, 7, 11, 12, 13, 14, 15, 19, 16, 2, 3] results in a Makespan of 1224
    Best found Solution.
    The permutation [8, 10, 17, 4, 9, 18, 1, 0, 5, 6, 7, 11, 12, 13, 14, 15, 19, 16, 2, 3] results in a Makespan of 1224

#### Aufgabe 3 - Implementierung II: Block-Move (4 Punkte)

Erweitern Sie den in der Lehrveranstaltung erstellten Solver um eine Block-Move Nachbarschaft für $k = 1$. Erstellen Sie die Klassen `BlockMoveK1` und `BlockNeighborhoodK1`.

**Hinweise**: 

Testen Sie die Nachbarschaft auf dem Datensatz `InputFlowshopSIST20.json`. Verwenden Sie die BestImprovement-Strategie und die FCFS-Startlösung. Mögliches erwartetes Ergebnis:


In [3]:
data = InputData("../../../data/Prescriptive_Homework/InputFlowshopSIST20.json")

localSearch = IterativeImprovement(data, 'BestImprovement', ['BlockK1'])

solver = Solver(data, 1008)

solver.RunLocalSearch(
    constructiveSolutionMethod='FCFS',
    algorithm=localSearch)

Generating an initial solution according to FCFS.
Constructive solution found.
The permutation [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1482
The permutation [0, 1, 2, 4, 5, 6, 7, 8, 9, 3, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1404
The permutation [17, 0, 1, 2, 4, 5, 6, 7, 8, 9, 3, 10, 11, 12, 13, 14, 15, 16, 18, 19] results in a Makespan of 1331
The permutation [17, 0, 1, 2, 4, 6, 7, 8, 9, 3, 10, 11, 12, 13, 14, 5, 15, 16, 18, 19] results in a Makespan of 1283
The permutation [17, 0, 1, 2, 6, 7, 8, 9, 4, 3, 10, 11, 12, 13, 14, 5, 15, 16, 18, 19] results in a Makespan of 1267
Best found Solution.
The permutation [17, 0, 1, 2, 6, 7, 8, 9, 4, 3, 10, 11, 12, 13, 14, 5, 15, 16, 18, 19] results in a Makespan of 1267


Habe blockSize direkt als Attribut in *BlockK3* Klasse integriert und setze es als default Wert in der Klasse *BlockNeighborhoodK3* auf 3. Anschließend kann ich beim aufrufen der *BlockNeighborhoodK3* Klasse das Attribut *blockSize* beliebig setzen. Somit kann ich einfach einen neuen Typ für BlockMoves innerhalb meiner *ImprovementAlgorithm.py* Datei integrieren, indem ich dort das Attribut *blockSize* beim Aufrufen der Klasse fix hinterlege.

Möglicher Output:

    Generating an initial solution according to FCFS.
    Constructive solution found.
    The permutation [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1482
    The permutation [0, 1, 2, 4, 5, 6, 7, 8, 9, 3, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1404
    The permutation [17, 0, 1, 2, 4, 5, 6, 7, 8, 9, 3, 10, 11, 12, 13, 14, 15, 16, 18, 19] results in a Makespan of 1331
    The permutation [17, 0, 1, 2, 4, 6, 7, 8, 9, 3, 10, 11, 12, 13, 14, 5, 15, 16, 18, 19] results in a Makespan of 1283
    The permutation [17, 0, 1, 2, 6, 7, 8, 9, 4, 3, 10, 11, 12, 13, 14, 5, 15, 16, 18, 19] results in a Makespan of 1267
    Best found Solution.
    The permutation [17, 0, 1, 2, 6, 7, 8, 9, 4, 3, 10, 11, 12, 13, 14, 5, 15, 16, 18, 19] results in a Makespan of 1267


**Tipp:**
<details>

Sein Sie clever statt tüchtig.
</details>